python/dgl/graphbolt/impl/ondisk_dataset.py

```diff
@@ -16,13 +32,183 @@
+        # Construct the data dict.
+        data_dict = {}
+        for edge_info in input_config["graph"]["edges"]:
+            edge_data = pd.read_csv(
+                dataset_path / edge_info["path"], names=["src", "dst"]
+            )
+            src = torch.tensor(edge_data["src"])
+            dst = torch.tensor(edge_data["dst"])
+            data_dict[tuple(edge_info["type"].split(":"))] = (src, dst)
+        # Construct the heterograph.
+        g = dgl.heterograph(data_dict, num_nodes_dict)
+
+    # 3. Load the node/edge features and add them to the sampling-graph.
+    if input_config["graph"].get("feature_data", None):
+        for graph_feature in input_config["graph"]["feature_data"]:
+            if graph_feature["domain"] == "node":
+                node_data = read_data(
+                    dataset_path / graph_feature["path"],
+                    graph_feature["format"],
+                    in_memory=graph_feature["in_memory"],
+                )
+                g.ndata[graph_feature["name"]] = node_data
+            if graph_feature["domain"] == "edge":
+                edge_data = read_data(
+                    dataset_path / graph_feature["path"],
+                    graph_feature["format"],
+                    in_memory=graph_feature["in_memory"],
+                )
+                g.edata[graph_feature["name"]] = edge_data
+
+    # 4. Convert the DGLGraph to a CSCSamplingGraph.
+    csc_sampling_graph = from_dglgraph(g)
+
+    # 5. Save the CSCSamplingGraph and modify the output_config.
+    output_config["graph_topology"] = {}
+    output_config["graph_topology"]["type"] = "CSCSamplingGraph"
+    output_config["graph_topology"]["path"] = (
+        processed_dir_prefix / "csc_sampling_graph.tar"
+    )
+
+    save_csc_sampling_graph(
+        csc_sampling_graph,
+        dataset_path / output_config["graph_topology"]["path"],
+    )
+    del output_config["graph"]
+
+    # 6. Load the node/edge features and do necessary conversion.
+    if input_config.get("feature_data", None):
+        for (feature, out_feature) in zip(
+            input_config["feature_data"], output_config["feature_data"]
+        ):
+            # Always save the feature in numpy format.
+            out_feature["format"] = "numpy"
+            out_feature["path"] = processed_dir_prefix / feature[
+                "path"
+            ].replace("pt", "npy")
+
+            if feature["format"] == "torch":
+                data = read_data(
+                    dataset_path / feature["path"],
+                    feature["format"],
+                    in_memory=feature["in_memory"],
+                )
+                save_data(
+                    data,
+                    dataset_path / out_feature["path"],
+                    out_feature["format"],
+                )
+            elif feature["format"] == "numpy":
+                # If the original format is numpy, just copy the file.
```

Comment:
Do we need to copy file? Can we just refer to the existing file to avoid unnecessary duplication?

Reply:
Same question to other unnecessary copies.

Reply:
I've already discussed it with @Rhett-Ying. We hope hat all the processed data can be placed in a unified folder.
I'll add a TODO here.